### PR TITLE
FIX: UOE While building Exists query for nested search_as_you_type field

### PR DESCRIPTION
**Pull request by Luci-MG**
### Description
FIX: Fixes UOE while building exists query for nested search_as_you_type field

### Related Issues
Resolves #[11821](https://github.com/opensearch-project/OpenSearch/issues/11821)

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] ~New functionality has been documented.~
  - [ ] ~New functionality has javadoc added~
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Signed-off-by: Mrudhul Guda <mg6.dev@gmail.com>

### CODE REVIEW

File: modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java [hunk 1]
```diff
@@ -41,15 +41,8 @@
...
-import org.apache.lucene.search.BooleanQuery;
-import org.apache.lucene.search.ConstantScoreQuery;
-import org.apache.lucene.search.DisjunctionMaxQuery;
-import org.apache.lucene.search.MatchNoDocsQuery;
-import org.apache.lucene.search.MultiPhraseQuery;
-import org.apache.lucene.search.Query;
-import org.apache.lucene.search.SynonymQuery;
-import org.apache.lucene.search.TermQuery;
+import org.apache.lucene.queryparser.classic.ParseException;
+import org.apache.lucene.search.*;
```
Comment (owaiskazi19) [inline modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java:45]: Let's remove the wildcard imports

File: modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java [hunk 2]
```diff
@@ -541,6 +532,34 @@ public void testMatchPhrase() throws IOException {
...
+        }));
+        QueryShardContext queryShardContext = createQueryShardContext(mapperService);
+        Query actual = new QueryStringQueryBuilder("field:*").toQuery(queryShardContext);
+        Query expected = new ConstantScoreQuery(new BooleanQuery.Builder()
+            .add(new NormsFieldExistsQuery("field.nested_field"), BooleanClause.Occur.SHOULD)
+            .add(new NormsFieldExistsQuery("field.nested_field._3gram"), BooleanClause.Occur.SHOULD)
+            .add(new NormsFieldExistsQuery("field.nested_field._2gram"), BooleanClause.Occur.SHOULD)
+            .add(new TermQuery(new Term("_field_names", "field.nested_field._index_prefix")), BooleanClause.Occur.SHOULD)
+            .build());
+        assertThat(actual, equalTo(expected));
```
Comment (owaiskazi19) [inline modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java:560]: Nit:
```suggestion
        assertEquals(expected, actual);
```

File: modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java [hunk 3]
```diff
@@ -64,10 +57,8 @@
 import org.opensearch.index.mapper.SearchAsYouTypeFieldMapper.SearchAsYouTypeFieldType;
 import org.opensearch.index.mapper.SearchAsYouTypeFieldMapper.ShingleFieldMapper;
 import org.opensearch.index.mapper.SearchAsYouTypeFieldMapper.ShingleFieldType;
-import org.opensearch.index.query.MatchPhrasePrefixQueryBuilder;
-import org.opensearch.index.query.MatchPhraseQueryBuilder;
-import org.opensearch.index.query.MultiMatchQueryBuilder;
-import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.index.query.*;
```
Comment (owaiskazi19) [inline modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java:60]: Here as well

**Review by owaiskazi19 — CHANGES_REQUESTED** [review]
LGTM with few comments

File: modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java (hunk 1 above)
Comment (Luci-MG) [reply modules/mapper-extras/src/test/java/org/opensearch/index/mapper/SearchAsYouTypeFieldMapperTests.java]: Sure!! sorry IDE auto optimized imports

**Comment by Luci-MG** [comment]
> Thanks @Luci-MG. You need to rebase with latest `main` to resolve the conflicts. LGTM with few comments.

Hello @owaiskazi19  Refactored and committed the changes based on the review requests, Please re-review again
Thank you :)

**Comment by Luci-MG** [comment]
> Thanks @Luci-MG. You need to rebase with latest `main` to resolve the conflicts. LGTM with few comments.

Hello @owaiskazi19 Refactored and committed the changes based on the review requests, Please re-review again
Thank you :)

**Comment by owaiskazi19** [comment]
@Luci-MG `:modules:mapper-extras:spotlessJavaCheck` failed. Can you run `./gradlew spotlessApply` and push again?

**Comment by Luci-MG** [comment]
> @Luci-MG `:modules:mapper-extras:spotlessJavaCheck` failed. Can you run `./gradlew spotlessApply` and push again?

Done this and pushed the changes

**Review by kkmr — APPROVED** [review]

**Review by owaiskazi19 — APPROVED** [review]
LGTM! @msfroh can you take another look? Thanks

**Comment by owaiskazi19** [comment]
@Luci-MG can you resolve the conflicts?

**Comment by Luci-MG** [comment]
Hello @owaiskazi19 Resolved the conflicts and rebased with the latest main

**Comment by msfroh** [comment]
> LGTM! @msfroh can you take another look? Thanks

The change looks nice and simple. 

I'm trying to figure out why the `existsQuery` threw an exception. What I've been able to piece together is that the SearchAsYouTypeFieldMapper was added in 2019, while the base `existsQuery` implementation was added in 2020. So that probably explains it.

Then I'm thinking, what path will this `existsQuery` take now? AFAIK, it will never match anything, looking at the super-class implementation:
```
        if (hasDocValues()) {
            return new DocValuesFieldExistsQuery(name());
        } else if (getTextSearchInfo().hasNorms()) {
            return new NormsFieldExistsQuery(name());
        } else {
            return new TermQuery(new Term(FieldNamesFieldMapper.NAME, name()));
        }
```

The prefix field doesn't have doc values or norms, so it will go into that third path. I could be mistaken, but I think `<field_name>._index_prefix` (the name of the prefix field) doesn't get added to the field names field, because it's not a "real" field.

**Comment by msfroh** [comment]
I think to me, the bigger question is **why** do we end up trying to call `existsQuery` on the prefix field type? I think there's a flaw in the logic somewhere else and the UOE is just highlighting that.

**Comment by msfroh** [comment]
> I think to me, the bigger question is **why** do we end up trying to call `existsQuery` on the prefix field type? I think there's a flaw in the logic somewhere else and the UOE is just highlighting that.

Okay -- I was able to answer that by debugging the unit test that you added. (Thank you very much for that, by the way!)

I now understand that the issue comes from the outer object field (named `field`). Because it does not exist in the mapper service as a direct type, it tries iterating through all fields whose name starts with `field.*`, which includes the "virtual" fields, like the prefix and shingle fields. So far as I can tell, we would run into the same problem with a `text` field with `index_prefixes` enabled.

 I tried the following change to `newObjectFieldExistsQuery` to avoid generating queries for those "virtual" subfields:
 
 ```
    private static Query newObjectFieldExistsQuery(QueryShardContext context, String objField) {
        BooleanQuery.Builder booleanQuery = new BooleanQuery.Builder();
        Collection<String> fields = context.simpleMatchToIndexNames(objField + ".*");
        for (String field : fields) {
            int dotPos = field.lastIndexOf('.');
            if (dotPos > 0 && field.charAt(dotPos + 1) == '_') { // This is a subfield (e.g. prefix) of a complex field type. Skip it.
                continue;
            }
            Query existsQuery = context.getMapperService().fieldType(field).existsQuery(context);
            booleanQuery.add(existsQuery, Occur.SHOULD);
        }
        return new ConstantScoreQuery(booleanQuery.build());
    }
```

**Comment by msfroh** [comment]
@Luci-MG  -- are you able to fix `newObjectFieldExistsQuery` instead? It looks like that's where the bug really resides.

**Comment by Luci-MG** [comment]
> @Luci-MG -- are you able to fix `newObjectFieldExistsQuery` instead? It looks like that's where the bug really resides.

Hello @msfroh let me check that will update you

**Comment by Luci-MG** [comment]
Hello @msfroh as you stated we can skip query generation for virtual sub-fields so made the changes accordingly 

1. In newObjectFieldExistsQuery skipped query generation by checking the virtual sub-fields
2. Modified expected case in the test case to skip virtual sub-fields like (_2gram, _3gram ...)
3. Re-Added throwing unsupportedexception for exists query (shall I remove it ?)

please review the changes and let me know if I need to make any additional changes

sorry for stalling MR was in the assumption that you have taken up the change for newObjectFieldExistsQuery

**Review by msfroh — APPROVED** [review]

**Pull request merged**